java/ql/src/experimental/Security/CWE/CWE-094/JexlInjectionLib.qll

```diff
@@ -0,0 +1,231 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+
+/**
+ * A taint-tracking configuration for unsafe user input
+ * that is used to construct and evaluate a Jexl expression.
+ * It supports both Jexl2 and Jexl3.
+ */
+class JexlInjectionConfig extends TaintTracking::Configuration {
+  JexlInjectionConfig() { this = "JexlInjectionConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof TaintedSpringRequestBody or
+    source instanceof RemoteFlowSource
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof JexlEvaluationSink }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node fromNode, DataFlow::Node toNode) {
+    any(TaintPropagatingJexlMethodCall c).taintFlow(fromNode, toNode) or
+    returnsDataFromBean(fromNode, toNode)
+  }
+}
+
+/**
+ * A data flow source for parameters that have
+ * a Spring framework annotation indicating remote user input from servlets.
+ */
+private class TaintedSpringRequestBody extends DataFlow::Node {
+  TaintedSpringRequestBody() {
+    exists(SpringServletInputAnnotation a | this.asParameter().getAnAnnotation() = a)
+  }
+}
+
+/**
+ * A sink for Expresssion Language injection vulnerabilities via Jexl,
+ * i.e. method calls that run evaluation of a Jexl expression.
+ */
+private class JexlEvaluationSink extends DataFlow::ExprNode {
+  JexlEvaluationSink() {
+    exists(MethodAccess ma, Method m, Expr taintFrom |
+      ma.getMethod() = m and taintFrom = this.asExpr()
+    |
+      m instanceof DirectJexlEvaluationMethod and ma.getQualifier() = taintFrom
+      or
+      m instanceof CallableCallMethod and ma.getQualifier() = taintFrom
```

Comment:
This is matching all calls to `java.util.concurrent.Callable` no matter if they are created from a JEXL script or not which is adding quite a few false positives

Reply:
To produce an alert, a `Callable` has to be tainted. This sink works together with `TaintPropagatingJexlMethodCall` that can make a `Callable` tainted if it is created from a tainted `JexlExpression` or `JexlScript`. There is no other way how a `Callable` can become tainted in this query. The sink is private, therefore it can't affect other queries. While running the query on multiple codebases, I didn't notice such a false positive. Please let me know if I am missing something. If you have a code snippet that results to a false positive, please let me know. I'll try to fix the query and add it as a test.

Reply:
Sure, you can use `apache/groovy` as test case. The project does not use `JEXL` and still gets 38 alerts. The problem arises when a `Callable` gets tainted. For example below `obj` is tainted and gets casted to groovy's `Closure` which implements the `Callable` interface:

```
Closure c = (Closure) ((Object[]) obj)[0];
..
c.call((Object)null);
```

It may be enough to set the sink in the `callable()` qualifier assuming that it will get executed at some point, maybe stored in an object field and then reached by a different flow.

Reply:
You're right, good catch!

> It may be enough to set the sink in the callable() qualifier assuming that it will get executed at some point, maybe stored in an object field and then reached by a different flow.

I've updated the query with your suggestion. Now it doesn't show false positives for Apache Groovy. Thanks!